### PR TITLE
DoBattleIntro state documentation

### LINKED ISSUE #4501: DoBattleIntro state documentation
### Description

"DoBattleIntro" currently does not have its various states documented and use raw numbers. This can make removing any of the battle intro states extremely tedious.

### Version

1.8.2 (Latest release)

### Upcoming/master Version

_No response_

### Discord contact info

_No response_

### PR DESCRIPTION
**Pull request by AsparagusEduardo**
## Description
Slight documentation to make it easier to modify intro states.
Also removed unused state.

## Issue(s) that this PR fixes
Fixes #4501

## **People who collaborated with me in this PR**
@ShinyDragonHunter

## **Discord contact info**
AsparagusEduardo

### CODE REVIEW

File: src/battle_main.c [hunk 1]
```diff
@@ -3421,6 +3421,30 @@ const u8* FaintClearSetData(u32 battler)
     return result;
 }
 
+enum
```
Comment (AlexOn1ine) [inline src/battle_main.c:3424]:
```suggestion
enum BattleIntroStates
```

**Comment by AsparagusEduardo** [comment]
Applied change

**Comment by AlexOn1ine** [comment]
correct me if I'm wrong but I think the variable declared on line [3452](https://github.com/rh-hideout/pokeemerald-expansion/pull/5231/files#diff-00f9630d2e74350419541de0421024862341855d15584748befb4366b201d114R3452) can be changed to `enum BattleIntroStates` instead of u8

**Comment by AsparagusEduardo** [comment]
> correct me if I'm wrong but I think the variable declared on line [3452](https://github.com/rh-hideout/pokeemerald-expansion/pull/5231/files#diff-00f9630d2e74350419541de0421024862341855d15584748befb4366b201d114R3452) can be changed to `enum BattleIntroStates` instead of u8

For that, it seems that I need to change `gBattleStruct->introState`'s type to `enum BattleIntroStates` too, and for that, move the enum to be next to `BattleStruct`'s definition.
Should we start doing this for enums used in structs?

**Comment by AlexOn1ine** [comment]
> > correct me if I'm wrong but I think the variable declared on line [3452](https://github.com/rh-hideout/pokeemerald-expansion/pull/5231/files#diff-00f9630d2e74350419541de0421024862341855d15584748befb4366b201d114R3452) can be changed to `enum BattleIntroStates` instead of u8
> 
> For that, it seems that I need to change `gBattleStruct->introState`'s type to `enum BattleIntroStates` too, and for that, move the enum to be next to `BattleStruct`'s definition. Should we start doing this for enums used in structs?

When you make the change in the Battle Struct could you check the space usage?

**Comment by AsparagusEduardo** [comment]
> When you make the change in the Battle Struct could you check the space usage in the struct?

If you mean EWRAM usage, no difference:
```
Memory region         Used Size  Region Size  %age Used
           EWRAM:      244554 B       256 KB     93.29%
           IWRAM:       30436 B        32 KB     92.88%
             ROM:    25696068 B        32 MB     76.58%
```
Weirdly enough, there's a 20-byte increase in *ROM* when setting the last enum to `300` or  `70000`.
```diff
    STATE_PRINT_PLAYER_2_SEND_OUT_TEXT,
-   STATE_SET_DEX_AND_BATTLE_VARS,
+   STATE_SET_DEX_AND_BATTLE_VARS = 300,
```
```diff
Memory region         Used Size  Region Size  %age Used
           EWRAM:      244554 B       256 KB     93.29%
           IWRAM:       30436 B        32 KB     92.88%
-            ROM:    25696068 B        32 MB     76.58%
+            ROM:    25696088 B        32 MB     76.58%
```

**Comment by AsparagusEduardo** [comment]
(I'll have to wait until I get home to check the actual struct size)

**Comment by AlexOn1ine** [comment]
> > When you make the change in the Battle Struct could you check the space usage in the struct?
> 
> If you mean EWRAM usage, no difference:
> 
> ```
> Memory region         Used Size  Region Size  %age Used
>            EWRAM:      244554 B       256 KB     93.29%
>            IWRAM:       30436 B        32 KB     92.88%
>              ROM:    25696068 B        32 MB     76.58%
> ```
> 
> Weirdly enough, there's a 20-byte increase in _ROM_ when setting the last enum to `300` or `70000`.
> 
> ```diff
>     STATE_PRINT_PLAYER_2_SEND_OUT_TEXT,
> -   STATE_SET_DEX_AND_BATTLE_VARS,
> +   STATE_SET_DEX_AND_BATTLE_VARS = 300,
> ```
> 
> ```diff
> Memory region         Used Size  Region Size  %age Used
>            EWRAM:      244554 B       256 KB     93.29%
>            IWRAM:       30436 B        32 KB     92.88%
> -            ROM:    25696068 B        32 MB     76.58%
> +            ROM:    25696088 B        32 MB     76.58%
> ```

no I meant `sizoof(struct BattleStruct)`

**Comment by ShinyDragonHunter** [comment]
A few things were missed btw. I'm not sure how to quote particular lines but...
"        if (!gBattleControllerExecFlags)
        {
            if (++gBattleCommunication[1] == gBattlersCount)
                (*state)++;
            else
                *state = 0;
        }
"
"
        if (B_FAST_INTRO == TRUE
          && !(gBattleTypeFlags & (BATTLE_TYPE_RECORDED | BATTLE_TYPE_RECORDED_LINK | BATTLE_TYPE_RECORDED_IS_MASTER | BATTLE_TYPE_LINK)))
            *state = 15; // Print at the same time as trainer sends out second mon.
        else
            (*state)++;
        break;
"

**Comment by AsparagusEduardo** [comment]
> no I meant `sizoof(struct BattleStruct)` but rom space is something tertu brought up before. It looks like changing the type of enum increases the rom size. See: #5106

Finally was able to check. The change increased the size of the struct from 1144b to 1152b

**Comment by AlexOn1ine** [comment]
> > no I meant `sizoof(struct BattleStruct)` but rom space is something tertu brought up before. It looks like changing the type of enum increases the rom size. See: #5106
> 
> Finally was able to check. The change increased the size of the struct from 1144b to 1152b

hm, then since both changes would lead to either increased rom size or increased BattleStruct size I'm leaning towards it being not important enough to justify the change.

**Comment by AsparagusEduardo** [comment]
> hm, then since both changes would lead to either increased rom size or increased BattleStruct size I'm leaning towards it being not important enough to justify the change.

You're talking about using an enum for `gBattleStruct->introState`, right? Or using an enum in general?

**Comment by AlexOn1ine** [comment]
> > hm, then since both changes would lead to either increased rom size or increased BattleStruct size I'm leaning towards it being not important enough to justify the change.
> 
> You're talking about using an enum for `gBattleStruct->introState`, right?

yes

**Review by AlexOn1ine — APPROVED** [review]

**Comment by AlexOn1ine** [comment]
> > no I meant `sizoof(struct BattleStruct)` but rom space is something tertu brought up before. It looks like changing the type of enum increases the rom size. See: #5106
> 
> Finally was able to check. The change increased the size of the struct from 1144b to 1152b

could you try limit it to 8 bits. will this prevent to increase the size? so `enum BattleInfoStates introState:8`?

**Comment by AsparagusEduardo** [comment]
> could you try limit it to 8 bits. will this prevent to increase the size? so `enum BattleInfoStates introState:8`?

Doesn't compile ^^;
image

**Comment by AsparagusEduardo** [comment]
I was able to remove the pointer reference alongside the bit limit change and it was kept as 1144b! :D

Since the enum will be in `include/battle.h`, should I rename the prefix from `STATE_` to `BATTLE_INTRO_STATE_`?

**Review by AlexOn1ine — APPROVED** [review]

**Pull request merged**